commons/src/main/kotlin/com/github/quillraven/commons/ashley/AbstractEntityFactory.kt

```diff
@@ -18,45 +18,49 @@ import ktx.log.debug
 import ktx.log.logger
 import kotlin.contracts.InvocationKind
 import kotlin.contracts.contract
-import kotlin.reflect.KClass
-import kotlin.reflect.full.createInstance
 
 /**
- * Class to create and cache [entity configurations][AbstractEntityConfiguration] of type [ConfigType].
+ * Class to create [entities][Entity] and create and cache
+ * [entity configurations][AbstractEntityConfiguration] of type [ConfigType].
  * It provides a DSL to support a type-safe creation of configuration objects.
  *
- * Use [config] to create, configure and add a configuration to the cache. The config id must be unique.
- * Otherwise a [GdxRuntimeException] is thrown.
+ * Use [configurations] and [config] to create, configure and add configurations to the cache.
+ * The config id must be unique. Otherwise a [GdxRuntimeException] is thrown.
  *
  * Use [get] to retrieve an already stored configuration. If the configuration is not existing then a
  * [GdxRuntimeException] is thrown.
  *
- * Use [newEntity] to create and add a new [Entity] to an [Engine] by using an already existing configuration.
+ * Use [newEntity] to create and add a new [Entity] to the [engine] by using an already existing configuration.
+ * If a [world] is specified then a [Box2DComponent] is initialized if needed.
+ *
+ * Use [configFactory] to define how a new instance of type [ConfigType] should be created within [config].
+ *
+ * Override [configureEntity] to add your own specific configurations after an entity was created.
  */
-class EntityConfigurations<ConfigType : AbstractEntityConfiguration>
-@PublishedApi
-internal constructor(
+abstract class AbstractEntityFactory<ConfigType : AbstractEntityConfiguration>(
+    val engine: Engine,
     @PublishedApi
-    internal val type: KClass<ConfigType>
+    internal val configFactory: () -> ConfigType,
+    val world: World? = null
 ) {
     @PublishedApi
-    internal val configurations = ObjectMap<String, ConfigType>()
+    internal val configCache = ObjectMap<String, ConfigType>()
 
     /**
-     * Creates and stores a new configuration with the given [id].
+     * Creates and stores a new configuration with the given [id] using the [configFactory] function.
      * Throws a [GdxRuntimeException] if the [id] is already existing.
      */
     inline fun config(id: String, block: ConfigType.() -> Unit) {
```

Comment:
Are the IDs finite and predefined? An enum might be cleaner here.

Reply:
refer to my explanation below. Because it is up to the user which configurations he will have, I cannot have an enum, correct?

However, the user can use an enum, like I do, and pass the enum name to this function. Could also be a number and then use enum ordinal but might be harder to debug/read the cache if needed.

Reply:
Yeah, you could define an empty interface that can be implemented by an enum, but it's not really an improvement in terms of usability.